File: packages/ergoscript/src/eval/collection.ts
/**
 * Collection arm — handles both `kind: 'Exprs'` and `kind: 'BoolConstants'`.
 *
 * Sigma-rust ref: `ergotree-interpreter/src/eval/collection.rs:22`
 *
 *     ctx.add_jit_cost(20)?;          // ConcreteCollection = Fixed(20)
 *     match self {
 *         Collection::BoolConstants(bools) => bools.clone().into(),
 *         Collection::Exprs { elem_tpe, items } => {
 *             let items_v = items.iter().map(|i| i.eval(env, ctx)).collect();
 *             // ... wrap in Value::Coll(CollKind::WrappedColl { ... }) or
 *             // Value::Coll(CollKind::NativeColl(NativeColl::CollByte(_)))
 *         }
 *     }
 *
 * Cost: ConcreteCollection = Fixed(20) (envelope) + sum of recursive item
 * costs. The envelope is charged BEFORE evaluating items (mirrors
 * sigma-rust); if a child item throws, the envelope cost has already been
 * added — same as the reference behavior.
 *
 * BoolConstants charges only the envelope: bools live inline in the variant
 * payload, no per-item recursion happens.
 *
 * Element-kind validation: for `Exprs`, every evaluated item is checked
 * against `elemTpe`. A mismatch is a static-type bug that the parser should
 * have caught — we surface it via `EvalError 'collection-elem-kind-mismatch'`
 * rather than silently building a malformed `Coll` SValue. sigma-rust
 * doesn't perform this check at eval time (it relies on the type checker
 * upstream), but we add it as a fail-fast guard for the verifier path.
 */

import type { Collection, SType, SValue } from '../mir/types'
import type { Env } from './env'
import type { EvalContext } from './eval-context'
import { EvalError } from './eval-context'
import { evalExpr } from './eval'
import { exprTpe } from '../mir/expr-tpe'
import { assertValueTypeSupported } from './_check-type'

export function evalCollection(e: Collection, env: Env, ctx: EvalContext): SValue {
  ctx.addCost(20)
  if (e.kind === 'BoolConstants') {
    // Each boolean constant is a BooleanConstant = FixedCost(5) on the JVM
    // (Constant.costKind, sigma `values.scala`). The `Exprs` arm below already
    // pays this via each Const eval; the packed BoolConstants form converts bits
    // directly without per-element eval, so charge 5*n here to match the JVM and
    // the Exprs arm (total = 20 + 5n). sigma-rust collection.rs, commit ee4c3c3f.
    ctx.addCost(5 * e.items.length)
    return {
      kind: 'Coll',
      elem: { tag: 'SBoolean' },
      items: e.items.map((b) => ({ kind: 'Boolean', value: b }) as SValue),
    }
  }
  // kind === 'Exprs'
  // ConcreteCollection.eval runs Value.checkType on each item (values.scala);
  // checkType is against the item's STATIC declared type, so a non-pair STuple
  // / non-unary SFunc declared type rejects (the JVM cannot represent such a
  // value). See eval/_check-type.ts.
  const items = e.items.map((item) => {
    const v = evalExpr(item, env, ctx)
    assertValueTypeSupported(exprTpe(item))
    return v
  })
  for (let i = 0; i < items.length; i++) {
    if (!kindMatchesType(items[i]!, e.elemTpe)) {
      throw new EvalError(
        `Collection.items[${i}] kind '${items[i]!.kind}' inconsistent with elemTpe '${e.elemTpe.tag}'`,
        'collection-elem-kind-mismatch'
      )
    }
  }
  return { kind: 'Coll', elem: e.elemTpe, items }
}

/**
 * Check that an evaluated item's value-kind matches the declared element
 * type. Only primitives are validated — composite types (`SColl`, `STuple`,
 * `SOption`, `SFunc`, etc.) and chain-state types (`SBox`, `SAvlTree`,
 * `SSigmaProp`, `SGroupElement`) are deferred to the future arm tasks
 * that introduce them. Returning `true` for unsupported tags is a
 * permissive default — it lets us land the primitive-coverage now without
 * blocking on phases 2c-2g.
 */
function kindMatchesType(v: SValue, t: SType): boolean {
  switch (t.tag) {
    case 'SBoolean':
      return v.kind === 'Boolean'
    case 'SByte':
      return v.kind === 'Byte'
    case 'SShort':
      return v.kind === 'Short'
    case 'SInt':
      return v.kind === 'Int'
    case 'SLong':
      return v.kind === 'Long'
    case 'SBigInt':
      return v.kind === 'BigInt'
    case 'SUnit':
      return v.kind === 'Unit'
    default:
      // Composite or chain-state types — defer to later phases.
      return true
  }
}

File: fixture-gen/src/cmds/ergoscript/eval/collection.rs
//! Collection arm — fixtures for both `kind: 'Exprs'` and `kind: 'BoolConstants'`.
//!
//! Sigma-rust ref: `ergotree-interpreter/src/eval/collection.rs:22`
//!     ctx.add_jit_cost(20)?;          // ConcreteCollection = Fixed(20)
//!     match self {
//!         Collection::BoolConstants(bools) => bools.clone().into(),
//!         Collection::Exprs { elem_tpe, items } => { /* eval each item */ }
//!     }
//!
//! Cost: ConcreteCollection = Fixed(20) (envelope) + recursive item costs.
//! Items are themselves `Const` here (cost 5 each), so a 3-element coll of
//! `Const` items costs 20 + 5 + 5 + 5 = 35; an empty coll costs 20.
//! `BoolConstants` charges only the envelope (20) — no per-item recursion,
//! since the bools are inlined in the variant.
//!
//! Uses test_util (gated by 'arbitrary' feature on ergotree-interpreter)
//! to drive the same evaluator the Scala node ships with — sigma-rust IS
//! the cost+value oracle for this arm.

use ergotree_interpreter::eval::test_util::try_eval_out;
use ergotree_ir::chain::context::Context;
use ergotree_ir::ergo_tree::{ErgoTree, ErgoTreeHeader};
use ergotree_ir::mir::collection::Collection;
use ergotree_ir::mir::expr::Expr;
use ergotree_ir::mir::value::Value;
use ergotree_ir::serialization::SigmaSerializable;
use ergotree_ir::types::stype::SType;
use serde_json::json;
use sigma_test_util::force_any_val;

use super::common::{value_to_json, EvalFixture, EvalFixtureFile};

pub fn generate() -> anyhow::Result<EvalFixtureFile> {
    let mut entries = Vec::new();

    // Case 1: BoolConstants kind — Coll[Boolean] of literals.
    // `Collection::from_bools` constructs the BoolConstants variant directly,
    // routing through the bool-packed wire opcode COLL_OF_BOOL_CONST. The TS
    // arm's BoolConstants branch must charge the envelope (20) and emit a
    // `{ kind: 'Coll', elem: { tag: 'SBoolean' }, items: Boolean[] }`.
    {
        let coll: Expr = Collection::from_bools(vec![true, false, true]).into();
        let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &coll)?;
        let bytes_hex = hex::encode(tree.sigma_serialize_bytes()?);
        let ctx = force_any_val::<Context>();
        let val: Value<'static> = try_eval_out(&tree.proposition()?, &ctx)?;
        entries.push(EvalFixture {
            name: "coll_bool_constants_3".to_string(),
            tree_bytes_hex: bytes_hex,
            opts_json: json!({}),
            expected_value_json: value_to_json(&val),
            expected_cost: ctx.jit_cost_value(),
        });
    }

    // Case 2: Exprs kind — Coll[Int] from Const exprs.
    // `Collection::new` with a non-bool `elem_tpe` always produces the
    // `Exprs` variant. The TS arm's Exprs branch must charge the envelope
    // (20) plus per-item recursive cost (5 each Const).
    {
        let items: Vec<Expr> = vec![
            Expr::Const(1i32.into()),
            Expr::Const(2i32.into()),
            Expr::Const(3i32.into()),
        ];
        let coll: Expr = Collection::new(SType::SInt, items)?.into();
        let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &coll)?;
        let bytes_hex = hex::encode(tree.sigma_serialize_bytes()?);
        let ctx = force_any_val::<Context>();
        let val: Value<'static> = try_eval_out(&tree.proposition()?, &ctx)?;
        entries.push(EvalFixture {
            name: "coll_exprs_int_3".to_string(),
            tree_bytes_hex: bytes_hex,
            opts_json: json!({}),
            expected_value_json: value_to_json(&val),
            expected_cost: ctx.jit_cost_value(),
        });
    }

    // Case 3: empty Coll[Long]. Verifies envelope-only cost (20) and empty
    // items array semantics (no item-kind validation triggered).
    {
        let coll: Expr = Collection::new(SType::SLong, vec![])?.into();
        let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &coll)?;
        let bytes_hex = hex::encode(tree.sigma_serialize_bytes()?);
        let ctx = force_any_val::<Context>();
        let val: Value<'static> = try_eval_out(&tree.proposition()?, &ctx)?;
        entries.push(EvalFixture {
            name: "coll_empty_long".to_string(),
            tree_bytes_hex: bytes_hex,
            opts_json: json!({}),
            expected_value_json: value_to_json(&val),
            expected_cost: ctx.jit_cost_value(),
        });
    }

    Ok(EvalFixtureFile {
        corpus: "eval_collection",
        entries,
    })
}
